### PR TITLE
[GSOC22] - D - Test the Three Way Merge UI

### PR DESCRIPTION
**Pull request by HoussemNasri**
Besides unit testing the three-way merge UI, the code was refactored for better testability

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by ThiloteE** [comment]
I don't know about the code, but at first glance, I have not seen any abnormal behaviour that would prevent merging, while trying this pr.

File: src/main/java/org/jabref/gui/mergeentries/newmergedialog/fieldsmerger/GroupMerger.java [hunk 1]
```diff
@@ -21,7 +21,7 @@ public String merge(String groupsA, String groupsB) {
         } else if (StringUtil.isBlank(groupsB)) {
             return groupsA;
         } else {
-            return Arrays.stream((groupsA + GROUPS_SEPARATOR + groupsB).split(GROUPS_SEPARATOR))
+            return Arrays.stream((groupsA + GROUPS_SEPARATOR + groupsB).split("\s*,\s*"))
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/mergeentries/newmergedialog/fieldsmerger/GroupMerger.java:24]: You should extract regexes whenever possible to instances of Pattern, e.g. Pattern.compile

Comment (HoussemNasri) [reply src/main/java/org/jabref/gui/mergeentries/newmergedialog/fieldsmerger/GroupMerger.java:24]: Ah I see, so it constructs the state machine only once

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/mergeentries/newmergedialog/fieldsmerger/GroupMerger.java]: For performance reasons, it doesn't need to create a new object everytime. 
Just use pattern.split()
 JDK11+ also has some fancy new methods 
 https://www.baeldung.com/java-regex-pre-compile

File: src/test/java/org/jabref/gui/mergeentries/FieldRowViewModelTest.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    /*
+// TODO: Fix this test
+    @Test
+    public void hasEqualLeftAndRightValuesShouldReturnTrueIfKeywordsAreEqual() {
+        var keywordsFieldViewModel = createViewModelForField(StandardField.KEYWORDS);
+        keywordsFieldViewModel.selectNonEmptyValue();
+        assertTrue(keywordsFieldViewModel.hasEqualLeftAndRightValues());
+    }*/
```
Comment (calixtus) [inline src/test/java/org/jabref/gui/mergeentries/FieldRowViewModelTest.java:117]: Anything planned how to continue on that?

**Review by calixtus — APPROVED** [review]
LGTM, just one question about a comment

File: src/test/java/org/jabref/gui/mergeentries/FieldRowViewModelTest.java (hunk 2 above)
Comment (HoussemNasri) [reply src/test/java/org/jabref/gui/mergeentries/FieldRowViewModelTest.java:117]: I think in the future we should have multiple field row implementations. e.g. GroupsFieldRow, KeywordFieldRow, etc. And in each implementation, we override a method to compare left and right values. At first, I thought that the differences between field rows are minimal thus we can abstract it away in one class and pass factories as needed like in the FieldMerger case. But now, I can think of many specific cases where a field has special requirements. It is very unintuitive to keep passing factories.

For the time being, I'm going to keep the comment as a reminder to implement a different comparison logic based on the given field.

Comment (Siedlerchr) [reply src/test/java/org/jabref/gui/mergeentries/FieldRowViewModelTest.java:117]: Okay then add the test, annotate it with @Disabled and add a small comment on it

**Review by calixtus — APPROVED** [review]

**Comment by HoussemNasri** [comment]
`ThemeManagerTest#installThemeOnScene` is failing. It throws a `java.lang.IllegalStateException: Called endChange before beginChange`. According to [this comment](https://bugs.openjdk.org/browse/JDK-8171894?focusedCommentId=14035995&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14035995), this exception is thrown when a list is modified from multiple threads. In our case, this list is the scene's stylesheets. I have never seen this exception while interacting with JabRef, nor while testing, so I think `testfx` might have caused it, but I need to dig more into it to confirm.

**Comment by calixtus** [comment]
It's hiccup if a more deeper issue with the stylesheet management of javafx and how the ThemeManager interacts with it. I don't think you should bother with it in this PR. As for now the tests are all green. So if @Siedlerchr is ok with this, we should merge this PR soon.

**Review by Siedlerchr — APPROVED** [review]

**Comment by Siedlerchr** [comment]
Yep, looks good so far. I would not spend time on this ThemeManager issue. As long as it works when running JabRef it's not worth spending time on it. Maybe we can disable the test on CI

**Pull request merged**